nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/marshallers/csv/DocumentModelListCSVWriter.java

```diff
@@ -18,46 +18,112 @@
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
 
+import org.apache.commons.csv.CSVPrinter;
+import org.apache.commons.lang3.StringUtils;
 import org.nuxeo.ecm.core.api.DocumentModel;
-import org.nuxeo.ecm.core.api.DocumentModelList;
-import org.nuxeo.ecm.core.io.registry.MarshallerRegistry;
 import org.nuxeo.ecm.core.io.registry.Writer;
 import org.nuxeo.ecm.core.io.registry.reflect.Setup;
-import org.nuxeo.ecm.core.schema.SchemaManager;
+import org.nuxeo.ecm.core.schema.types.Field;
+import org.nuxeo.ecm.core.schema.types.Schema;
 
 /**
  * @since 10.3
  */
 @Setup(mode = SINGLETON, priority = REFERENCE)
-public class DocumentModelListCSVWriter extends AbstractCSVWriter<DocumentModelList> {
+public class DocumentModelListCSVWriter extends AbstractCSVWriter<List<DocumentModel>> {
 
     public DocumentModelListCSVWriter() {
-        super(DocumentModelList.class);
+        super();
     }
 
-    @Inject
-    protected SchemaManager schemaManager;
-
-    @Inject
-    protected MarshallerRegistry registry;
-
     @Override
-    public void write(DocumentModelList entity, Class<?> clazz, Type genericType, MediaType mediatype, OutputStream out)
-            throws IOException {
-        Writer<DocumentModel> writer = registry.getWriter(null, DocumentModel.class, TEXT_CSV_TYPE);
+    protected void write(List<DocumentModel> entity, CSVPrinter printer) throws IOException {
+        Writer<DocumentModel> writer = registry.getWriter(ctx, DocumentModel.class, TEXT_CSV_TYPE);
         for (DocumentModel doc : entity) {
-            writer.write(doc, DocumentModel.class, null, TEXT_CSV_TYPE, out);
-            out.write("\n".getBytes(UTF8));
+            writer.write(doc, DocumentModel.class, DocumentModel.class, TEXT_CSV_TYPE,
+                    new OutputStreamWithCSVWriter(printer));
+            printer.println();
+        }
+    }
+
+    @Override
+    protected void writeHeader(List<DocumentModel> entity, CSVPrinter printer) throws IOException {
+        writeDocumentModelListHeader(entity, printer);
+        printer.println();
+    }
+
+    protected void writeDocumentModelListHeader(List<DocumentModel> list, CSVPrinter printer) throws IOException {
+        printer.print(REPOSITORY_FIELD);
+        printer.print(UID_FIELD);
+        printer.print(PATH_FIELD);
+        printer.print(TYPE_FIELD);
+        printer.print(STATE_FIELD);
+        printer.print(PARENT_REF_FIELD);
+        printer.print(IS_CHECKED_OUT_FIELD);
+        printer.print(IS_VERSION_FIELD);
+        printer.print(IS_PROXY_FIELD);
+        printer.print(PROXY_TARGET_ID_FIELD);
+        printer.print(VERSIONABLE_ID_FIELD);
+        printer.print(CHANGE_TOKEN_FIELD);
+        printer.print(IS_TRASHED_FIELD);
+        printer.print(TITLE_FIELD);
+        printer.print(VERSION_LABEL_FIELD);
+        printer.print(LOCK_OWNER_FIELD);
+        printer.print(LOCK_CREATED_FIELD);
+        printer.print(LAST_MODIFIED_FIELD);
```

Comment:
There's similar code for the whole method in `DocumentModelCSVWriter`, it should be factored.